ui/contexts/snaps/snap-interface.tsx

```diff
@@ -0,0 +1,166 @@
+import {
+  FormState,
+  InterfaceState,
+  UserInputEventType,
+} from '@metamask/snaps-sdk';
+import { debounce } from 'lodash';
+import React, {
+  FunctionComponent,
+  createContext,
+  useContext,
+  useEffect,
+  useRef,
+} from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { getMemoizedInterface } from '../../selectors';
+import { handleSnapRequest, updateInterfaceState } from '../../store/actions';
+import { mergeValue } from './utils';
+
+export type HandleEvent = (event: UserInputEventType, name?: string) => void;
+
+export type HandleInputChange = (
+  name: string,
+  value: string | null,
+  form?: string,
+) => void;
+
+export type GetValue = (name: string, form?: string) => string | undefined;
+
+export type SnapInterfaceContextType = {
+  handleEvent: HandleEvent;
+  getValue: GetValue;
+  handleInputChange: HandleInputChange;
+};
+
+export const SnapInterfaceContext =
+  createContext<SnapInterfaceContextType | null>(null);
+
+export type SnapInterfaceContextProviderProps = {
+  interfaceId: string;
+  snapId: string;
+};
+
+/**
+ * The Snap interface context provider that handles all the interface state operations.
+ *
+ * @param params - The context provider params.
+ * @param params.children - The childrens to wrap with the context provider.
+ * @param params.interfaceId - The interface ID to use.
+ * @param params.snapId - The Snap ID that requested the interface.
+ * @returns The context provider.
+ */
+export const SnapInterfaceContextProvider: FunctionComponent<
+  SnapInterfaceContextProviderProps
+> = ({ children, interfaceId, snapId }) => {
+  const dispatch = useDispatch();
+  const { state: initialState } = useSelector(
+    (state) => getMemoizedInterface(state, interfaceId),
+    // Prevents the selector update.
+    // We do this to avoid useless re-renders.
+    () => true,
+  );
+
+  // We keep an internal copy of the state to speed-up the state update in the UI.
+  // It's kept in a ref to avoid useless re-rendering of the entire tree of components.
+  const internalState = useRef<InterfaceState>(initialState ?? {});
+
+  // Since the internal state is kept in a reference, it won't update when the interface is updated.
+  // We have to manually update it
+  useEffect(() => {
+    internalState.current = initialState;
+  }, [initialState]);
+
+  // The submittion of user input events is debounced to avoid crashing the snap if
+  // there's too much events sent at the same time
+  const snapRequestDebounced: HandleEvent = debounce(
```

Comment:
Do we still need to debounce this when we only have button clicks and form submit?

Reply:
Yes otherwise you can crash the snap by spamming the button